docs/book/src/clusterctl/provider-contract.md

```diff
@@ -241,6 +245,40 @@ Templates writers should use the common variables to ensure consistency across p
 Additionally, the value of the command argument to `clusterctl generate cluster <cluster-name>` (`<cluster-name>` in this case), will
 be applied to every occurrence of the `${ CLUSTER_NAME }` variable.
 
+### ClusterClass definitions
+
+An infrastructure provider could publish a **cluster class definition** file to be used by `clusterctl generate cluster` that will be used along
+with the workload cluster templates.
+This is a single YAML with _all_ the objects required that make up the cluster class.
+
+The following rules apply:
+
+#### Naming conventions
+
+ClusterClass definitions MUST be stored in the same folder as the component YAML and follow this naming convention:
+1. The cluster class definition should be names `clusterclass-{class-name}.yaml`, e.g `clusterclass-prod.yaml`
+
+`{class-name}` is the name of ClusterClass. 
+
+Each provider should create user facing documentation with the list of available cluster class definitions.
+
+#### Target namespace
+
+The cluster class definition YAML MUST asume the target namespace already exists.
+
+All objects in the cluster class definition YAML MUST be deploymned in the same namespace.
```

Comment:
```suggestion
All objects in the cluster class definition YAML MUST be deployed in the same namespace.
```
I wonder if we should recommend that the namespace is not set.

What would happen if someone sets a specific namespace? Would it be overwritten automatically or only when `${NAMESPACE}` is used?

Reply:
Maybe, yup. But not if it is set in the refs in the ClusterClass.

Reply:
Given that we don't want to recommend using variables in ClusterClass we should probably recommend that the namespace should not be set at all.

So that then the combination of fixTargetNamespace and our defaulting webhook will make sure the right namespace is used. Otherwise things can get messy if the top-level namespace is overwritten and is not the same as the one hard-coded in the refs.

Reply:
>What would happen if someone sets a specific namespace? Would it be overwritten automatically or only when ${NAMESPACE} is used?

It will be overridden automatically. 

As long as we have a note saying that none of the references in the ClusterClass set the namespace things should work as expected. (Will add this note in the contract.)

>Given that we don't want to recommend using variables in ClusterClass we should probably recommend that the namespace should not be set at all.

We can definitely add this recommendation to make things consistent. But just FYI, it is totally fine if a namespace is specified in the resource. As long as none of the references use a namespace it will work.

Reply:
Got it, thx